lib/charms/loki_k8s/v0/loki.py

```diff
@@ -54,24 +213,45 @@ def __init__(self, charm, relation_name: str):
         super().__init__(charm, relation_name)
         self.charm = charm
         self._relation_name = relation_name
+        self.container = self.charm.unit.get_container("loki")
         events = self.charm.on[relation_name]
         self.framework.observe(events.relation_changed, self._on_logging_relation_changed)
+        self.framework.observe(events.relation_departed, self._on_logging_relation_departed)
 
-    ##############################################
-    #               RELATIONS                    #
-    ##############################################
     def _on_logging_relation_changed(self, event):
-        """Anytime there are changes in relations between Loki
-        provider and consumer charms the Loki charm is informed,
-        through a `RelationChanged` event.
-        The Loki charm then updates relation data with the Loki Push API url.
+        """Handle changes in related consumers.
+
+        Anytime there are changes in relations between Loki
+        and its consumers charms, Loki set the `loki_push_api`
+        into the relation data.
+        Besides Loki generates alert rules files based what
+        consumer charms forwards,
+
+        Args:
+            event: a `CharmEvent` in response to which the Loki
+                charm must update its relation data.
+        """
+
+        if event.relation.data[self.charm.unit].get("data") is None:
+            event.relation.data[self.charm.unit]["data"] = self._loki_push_api
+            logger.debug("Saving Loki url in relation data %s", self._loki_push_api)
+
+        if event.relation.data.get(event.relation.app) is not None:
```

Comment:
Consider documenting rationale.
Should probably be more specific with the `.get`: might add unrelated things to app data in the future that should/shouldn't trigger this.